SYCL/InvokeSimd/invoke_simd_conv.cpp

```diff
@@ -0,0 +1,184 @@
+  return calc(val);
+}
+
+class ESIMDSelector : public device_selector {
+  // Require GPU device unless HOST is requested in SYCL_DEVICE_FILTER env
+  virtual int operator()(const device &device) const {
+    if (const char *dev_filter = getenv("SYCL_DEVICE_FILTER")) {
+      std::string filter_string(dev_filter);
+      if (filter_string.find("gpu") != std::string::npos)
+        return device.is_gpu() ? 1000 : -1;
+      if (filter_string.find("host") != std::string::npos)
+        return device.is_host() ? 1000 : -1;
+      std::cerr
+          << "Supported 'SYCL_DEVICE_FILTER' env var values are 'gpu' and "
+             "'host', '"
+          << filter_string << "' does not contain such substrings.\n";
+      return -1;
+    }
+    // If "SYCL_DEVICE_FILTER" not defined, only allow gpu device
+    return device.is_gpu() ? 1000 : -1;
+  }
+};
+
+inline auto createExceptionHandler() {
+  return [](exception_list l) {
+    for (auto ep : l) {
+      try {
+        std::rethrow_exception(ep);
+      } catch (sycl::exception &e0) {
+        std::cout << "sycl::exception: " << e0.what() << std::endl;
+      } catch (std::exception &e) {
+        std::cout << "std::exception: " << e.what() << std::endl;
+      } catch (...) {
+        std::cout << "generic exception\n";
+      }
+    }
+  };
+}
+
+template <class, class, bool> class TestID;
+
+template <class SpmdT, class SimdElemT, bool IsUniform> bool test(queue q) {
+  // 3 subgroups per workgroup
+  unsigned GroupSize = VL * 3;
+  unsigned NGroups = 7;
+  unsigned Size = GroupSize * NGroups;
+  SimdElemT *A = malloc_shared<SimdElemT>(Size, q);
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = (SimdElemT)i;
+  }
+  sycl::range<1> GlobalRange{Size};
+  sycl::range<1> LocalRange{GroupSize};
+  sycl::nd_range<1> Range(GlobalRange, LocalRange);
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<TestID<SpmdT, SimdElemT, IsUniform>>(
+          Range, [=](nd_item<1> ndi) [[intel::reqd_sub_group_size(VL)]] {
+            sub_group sg = ndi.get_sub_group();
+            SpmdT val = (SpmdT)sg.get_group_linear_id(); // 0 .. GroupSize-1
+            SimdElemT res = 0;
+
+            if constexpr (IsUniform) {
+              res =
+                  invoke_simd(sg, SIMD_CALLEE_UNIFORM<SimdElemT>, uniform{val});
+            } else {
+              res = invoke_simd(sg, SIMD_CALLEE<SimdElemT>, val);
+            }
+            uint32_t i = ndi.get_global_linear_id();
+            A[i] = res;
+          });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    sycl::free(A, q);
+    return false;
+  }
+  int err_cnt = 0;
+
+  for (unsigned i = 0; i < Size; ++i) {
+    unsigned group_id = i / GroupSize;
+    uint32_t sg_id = (i - (group_id * GroupSize)) / VL;
+    SimdElemT test = A[i];
+    SimdElemT gold = calc((SimdElemT)sg_id);
+    if (test != gold) {
+      if (++err_cnt < 10) {
```

Comment:
Nit but these might as well be merged into one if-statement to save an indention level.

Reply:
Thanks, updated.